### PR TITLE
Document and test `Base.rename`

### LINKED ISSUE #41584: Add atomicity keyword arg to `mv`
`mv` is often specified to be atomic on file systems, i.e. programs opening
a file will either get a handle to the old file or the new file, but never see an
incomplete file in the middle. This is highly useful for live replacing configuration
files, etc. Our `mv` is best-effort atomic, but if the files to be moved are on
different file systems, this can have problematic consequences. For example,
if the destination filesystem is full, it is possible for the destination file to
get truncated at arbitrary points. It would be better to have an option for
`mv` to automatically fail if atomicity cannot be guaranteed (maybe that
should even be the default).

### PR DESCRIPTION
**Pull request by nhz2**
Follow up of #55384

### CODE REVIEW

**Comment by LilithHafner** [comment]
Thanks for the docs and tests! Those should be easy to merge :)

For public API, should this be a method/kwarg of the existing `mv`? When we have multiple names doing similar things, it's important for the difference in names to reflect the difference in behavior. It's not intuitive to me that `mv` would be less atomic than `replace`. It's also not clear from the docstrings what, exactly, the semantic differences are between these functions.

**Comment by nhz2** [comment]
I'm using the function name `rename` because this function is basically [rename(2)](https://man7.org/linux/man-pages/man2/rename.2.html) using libuv to provide an approximation of `rename` on Windows.

I would say the main difference between `rename` and `mv` is `mv` prioritizes being consistent across different filesystems even if that means having TOCTTOU issues, not being `CTRL-C` safe, or deleting `dst` before copying `src`.
On the other hand `rename` prioritizes being atomic if possible, or otherwise ensuring external processes looking at an existing `newpath` should see either the contents of the previous `newpath` or `oldpath`, but never a missing or partially written file/directory, even if that means behavior will be inconsistent between Windows and Unix.

There is some discussion on this in https://github.com/microsoft/STL/pull/2062

**Comment by nhz2** [comment]
Here are some examples of where this functionality is used currently.

https://github.com/JuliaIO/HDF5.jl/blob/fbcd95dde2083c9f8c8ef6defaaf7fafd2552123/src/file.jl#L155

https://github.com/yha/AtomicFileWrite.jl/blob/0fbb7cf5b0e8ef899b2682ea70f5483923636504/src/AtomicFileWrite.jl#L24

https://github.com/medyan-dev/MEDYANSimRunner.jl/blob/67d5b42cc599670486d5d640260a95e951091f7a/src/file-saving.jl#L83

https://github.com/JuliaLang/Pkg.jl/pull/4001

**Comment by LilithHafner** [comment]
Folks using something does not make it part of the public API.

Widespread dependence on internals means we should have some API to expose those internals, but does not necessarily mean we should publicize the internals as is.

Tagging triage to bikeshed the name we should give the functionality that currently lives in the internal `Base.Filesystem.rename` function.

File: base/file.jl [hunk 1]
```diff
@@ -1177,16 +1177,28 @@ function unlink(p::AbstractString)
...
+Specifically, currently on Windows:
+1. `rename` will fail if `oldpath` or `newpath` are opened files.
+2. `rename` will fail if `newpath` is an existing directory.
+3. `rename` may work if `newpath` is a file and `oldpath` is a directory.
+4. `rename` may remove `oldpath` if it is a hardlink to `newpath`.
+
 See also: [`mv`](@ref).
+
+!!! compat "Julia 1.12"
+    This method was added in Julia 1.12.
```
Comment (LilithHafner) [inline base/file.jl:1201]: This was not added in 1.12 (even if we publicize it in 1.12)

**Review by LilithHafner — COMMENTED** [review]
Do we already follow the POSIX standard for `mv`?

Can we remove publicizing this and adding it to the manual and merge the docstring and tests now? Then we can address publicizing in a separate PR.

**Comment by LilithHafner** [comment]
From Triage: Thank you so much for adding docs and tests. None of us on triage know filesystem details well enough to say for sure weather this API is appropriate.

If it's not too hard, motivating examples for new API are always appropriate and helpful.

File: base/file.jl (hunk 1 above)
Comment (nhz2) [inline base/file.jl:1201]:
```suggestion
    This method was publicized in Julia 1.12.
```

Comment (nhz2) [reply base/file.jl:1201]: The `rename` function in 1.11 had a different behavior. It was changed in #55384

**Comment by LilithHafner** [comment]
Thanks!

**Comment by LilithHafner** [comment]
Feel free to open a followup to publicize

**Pull request merged**